dialogue-blocking-channels/src/main/java/com/palantir/dialogue/blocking/BlockingChannelAdapter.java

```diff
@@ -52,42 +57,66 @@ private BlockingChannelAdapter() {}
     private static final class BlockingChannelAdapterChannel implements Channel {
 
         private final BlockingChannel delegate;
-        private final ListeningExecutorService executor;
+        private final ExecutorService executor;
 
-        BlockingChannelAdapterChannel(BlockingChannel delegate, ListeningExecutorService executor) {
+        BlockingChannelAdapterChannel(BlockingChannel delegate, ExecutorService executor) {
             this.delegate = delegate;
             this.executor = executor;
         }
 
         @Override
         public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
-            return executor.submit(new BlockingChannelAdapterTask(delegate, endpoint, request));
+            SettableFuture<Response> result = SettableFuture.create();
+            BlockingChannelAdapterTask runnable = new BlockingChannelAdapterTask(delegate, endpoint, request, result);
+            Future<?> future = executor.submit(runnable);
+            // The executor task should be interrupted on termination
+            result.addListener(() -> future.cancel(true), MoreExecutors.directExecutor());
```

Comment:
seems a tiny bit odd that this fires even when the result is completed normally... why not use `Futures.addCallback` to only fire in the event that you get a CancellationException?

Reply:
it's a bit heavier weight, but probably not something to optimize around at this point, updated.